Mathlib/RingTheory/Valuation/ValuationSubring.lean

```diff
@@ -416,6 +416,7 @@ variable {Γ Γ₁ Γ₂ : Type*} [LinearOrderedCommGroupWithZero Γ]
   (v₁ : Valuation K Γ₁) (v₂ : Valuation K Γ₂)
 
 /-- The valuation subring associated to a valuation. -/
+@[reducible]
```

Comment:
Is there a reason why this is necessary?

Reply:
My consideration here is to make instances equipped on `Valuation.integer` adopted to `Valuation.valuationSubring` automatically. Otherwise, these instances have to be written twice. (Similar things will happen for `Valued.integer` and `Valued.valuationSubring`, e.g. there will be different instances `Algebra vR.v.integer vA.v.integer` and `Algebra vR.v.valuationSubring vA.v.valuationSubring`) I doubt there will be any instances that should only work on `Valuation.integer` but not `Valuation.valuationSubring`.

Reply:
Do you really need `Algebra vR.v.valuationSubring vA.v.valuationSubring` though? The main thing I'm concerning is about the performance impact.

Reply:
A few weeks ago we just changed the definition of `ringOfIntegers` from a `Subring` to a `Type*` because of performance reasons and I'm afraid this is going the opposite direction.

Reply:
> Do you really need `Algebra vR.v.valuationSubring vA.v.valuationSubring` though? The main thing I'm concerning is about the performance impact.

Yes, I need it to create the `IsLocalRingHom vR.v.valuationSubring vA.v.valuationSubring` instance, when `R` and `A` are fields. So should I indeed write all these instances twice? That's acceptable.

Reply:
I'm not sure I understand what `IsLocalRingHom vR.v.valuationSubring vA.v.valuationSubring` means

Reply:
Sorry, it should be`IsLocalRingHom (algebraMap vR.v.valuationSubring vA.v.valuationSubring)`.

Reply:
Why not `IsLocalRingHom (algebraMap vR.v.integer vA.v.integer)` then? And maybe put `LocalRing` instances on `vR.v.integer` instead.

Reply:
Well, if we only use `vR.v.integer` instead of `vR.v.valuationSubring` in instances, mathematically you simply lose some information when `R` and `A` are fields. For example, in some propositions, one really needs that `vR.v.integer` is a valuation subring of `R` and ` IsScalarTower vR.v.integer vA.v.integer A` together. It would be inconvenient to do this without `IsScalarTower vR.v.valuationSubring vA.v.valuationSubring A`.

Reply:
Hmmm. Do you have an example? Mathematically $A \hookrightarrow \mathrm{Frac}(A)$ with $A$ being a valuation ring should be everything `valuationSubring` entails.

Reply:
> Hmmm. Do you have an example? Mathematically A↪Frac(A) with A being a valuation ring should be everything `valuationSubring` entails.

Yes, you are right. What I need is exactly $$A \hookrightarrow \text{Frac}(A).$$ But for general `R`, we do not have `R = Frac vR.v.integer` unless `R` is a field. So do you think I should write instances like`LocalRing vR.v.integer` and `IsFractionRing vR.v.integer R` when `R` is a field (they fail when `R` is not a field)? If I use `vR.v.valuationSubring`, they can be inferred automatically.

Reply:
Yes I'd prefer `ValuationRing vR.v.integer` and `IsFractionRing vR.v.integer R` over putting more structure on `vR.valuationSubring` in case we would like to make `Valuation.integer` a type in the future due to performance concerns.

Reply:
I see, I'll work on it. Thank you for your suggestion! I learned a lot : )